### PR TITLE
Setup registry

### PR DESCRIPTION
**Pull request by grst**
To build the registry,
 * I defined a `schema.json` with [json-schema](https://json-schema.org/). This is nice because it is self-documenting, very explicit and we get validation for free via pre-commit. 
 * I added example entries in the `packages.yml` file. 

The fields and fieldnames are inspired by MDAKit as @ivirshup suggested. For now it's the absolute minimum of tags we can get away with, we can discuss what to add in addition. In general, I'm leaning towards avoiding to adding information that is available elsewhere (e.g. the pyproject.toml of each project), as I'm afraid it will get out-of-sync at some point.

### CODE REVIEW

File: schema.json [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+                        "pypi": {
+                            "description": "PyPI identifier",
+                            "type": "string"
+                        },
+                        "conda": {
+                            "description": "Conda identifier, including channel (e.g. conda-forge::scanpy)",
+                            "type": "string",
+                            "pattern": "(.*)::(.*)"
+                        }
+                    }
```
Comment (grst) [inline schema.json:46]: For now only PyPI or conda. We can add more entries here as we start adding our first Julia/CLI/... package

**Comment by grst** [comment]
More fields we could discuss adding (please comment if you think we should add anything else): 

- publications
- authors
- maintainers
- codecov

I have mixed feelings about the first three, as I'm afraid nobody would ever bother to update them.

**Review by ivirshup — COMMENTED** [review]
Looks great! Definitely would have gone with json-schema too.

Some thoughts:

* Could packages get their own file? I'm worried about conflicts for one long manifest.
* I would eventually like to be able to run tests for ecosystem packages. What do you think of including fields for installing + running tests?

**Comment by grst** [comment]
> Could packages get their own file? I'm worried about conflicts for one long manifest.

I was expecting you to ask this :joy: 

What kind of conflicts? I can see points for both, but my arguments in favor of a single file were: 
 * no need to decide about filenames. Can't always be the same as the package name if the package name is weird. 
 * if we end up with 1000 ecosystem packages it's a lot. Navigating 10k lines is no problem with a modern editor. Having 1k files in a single directory would already slow down a `ls` on our cluster though...
 * if it's in a single file, its easier to download it via API calls. It would even be an option to load it dynamically from our website without any additional tooling.

**Comment by ivirshup** [comment]
Re fields we could add:

* I think publication(s?) as an optional field would be good. Make it easier to find the citation.
* I think authors/ maintainers should be required. I think name + github handle would be good. At the very least, so we know who to get in touch with.
* I would also like coverage as an optional field. Not completely sure how to include it though, since people may use different CI services.

**Comment by ivirshup** [comment]
Worried this will get a little interleaved but:

## Multi vs single file

> What kind of conflicts

* Two PRs open at the same time
* Update some entries, while a PR is open
* Someone accidentally runs formatting

> no need to decide about filenames. Can't always be the same as the package name if the package name is weird.

True, but maybe not so bad? I would think any distributable package name should have some constraints (from the language/ package manager) on it.

> if we end up with 1000 ecosystem packages it's a lot.

I don't think we need to worry about the cluster use case, why would we need this on a cluster?

Also, I think we would need to allow having a different file name than package name. For packages that exist in multiple languages, for example.

> if it's in a single file, its easier to download it via API calls. It would even be an option to load it dynamically from our website without any additional tooling.

I think this is pretty solvable. A bundle could be created for each commit to main branch, or you could just read directly from the main branch archive.

What if you only wanted the metadata for a single package? Maybe to render a selection of projects, or run tests for one?

Also, in future, maybe we don't even keep a list. Maybe the file lives on each packages repo, and we just keep a list of repos here (in this case, we wouldn't need to be responsible for merging updates either). Could even be generated by the cookie cutter template.

-------------

Overall, I think it's more common to keep files separate. For example how bioconda/ conda-forge do it.

**Comment by grst** [comment]
> > Two PRs open at the same time
> Update some entries, while a PR is open
> Someone accidentally runs formatting

Maybe not so bad with a prettier pre-commit action, but agreed that there might still be some problems. 

> I think this is pretty solvable. 

Of course this is solvable. It just requires additional steps.

> What if you only wanted the metadata for a single package? Maybe to render a selection of projects, or run tests for one?

I'd argue that the overhead of parsing a 10k line yaml file is negligible.

> Also, in future, maybe we don't even keep a list. Maybe the file lives on each packages repo, and we just keep a list of repos here (in this case, we wouldn't need to be responsible for merging updates either). Could even be generated by the cookie cutter template.

this is a very interesting idea! Could be a `.scverse.yml` in the cookiecutter template. 

So what are you suggesting? Something like
```
.
└── packages
    ├── decoupler.yml
    └── infercnvpy.yml
```

In the MDAKit and conda they have an additional folder for each package and then just a `meta.yml` file in it. For conda this makes kind of sense as there might also be a `build.sh` or patch files in that folder. I don't think we'd need that?

**Comment by grst** [comment]
> I would eventually like to be able to run tests for ecosystem packages. What do you think of including fields for installing + running tests?

In principle I'm not opposed, but I'm afraid that the associated tooling takes too much time to implement for now? If we have this field we'd probably also want to ensure that the commands work? 

---

programming language (or languages as a list) might be another field we could add.

**Comment by ivirshup** [comment]
> So what are you suggesting? Something like:

Yeah, like that.

> an additional folder for each package and then just a meta.yml file in it. For conda this makes kind of sense ... I don't think we'd need that?

I don't think we'd need it, but I'm mildly for it. It would be forward looking, since maybe we'd want more files in the future. One use case could be including a logo thumbnail.

> > What if you only wanted the metadata for a single package? Maybe to render a selection of projects, or run tests for one?
> I'd argue that the overhead of parsing a 10k line yaml file is negligible.

Downloading it would be more expensive though. And the file hash would change for a change to any package, so you couldn't tell which entries had been updated.

Other use case is versioning of this schema. I think it's going to be easier to do validation on a single file using a single version of the schema than on a list of entries possibly using different versions.

Also I think we should probably version the schema.

**Comment by grst** [comment]
Ok, so I went with the conda-forge style. Having icons is maybe not a bad idea in the near future. 

Regarding versioning: there doesn't seem to be a standard way for versioning json schemata. IMO, as long as all `meta.yaml` files are within this repo they should always be updated together with `schema.json` and validate against the most recent version. This would change if we allow yml files in other repos. 

Regarding additional fields, here's again an overview of what we discussed: 

 * publications (list of DOIs, optional)
 * authors/maintainers (list of name/email/github, required) -> even if not kept up to date this serves as a point of contact for us
 * coverage (link to web service computing test coverage, optional)
 * test_command (shell command to install package and execute tests within the repository, mandatory)

File: packages/decoupler/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    Python package to infer biological activities from omics data using a collection of methods.
+project_home: https://github.com/saezlab/decoupler-py
+documentation_home: https://decoupler-py.readthedocs.io/en/latest/
+tutorials_home: https://decoupler-py.readthedocs.io/en/latest/
+install:
+    pypi: decoupler
+tags:
+    - TF
+    - GSEA
+    - Pathway
```
Comment (ivirshup) [inline packages/decoupler/meta.yaml:12]: Should the tags be acronyms?

Comment (grst) [reply packages/decoupler/meta.yaml:12]: Until we define a controlled vocabulary, these are anyway just arbitrary examples showcasing that one can add tags to the yaml file :shrug:

Comment (ivirshup) [reply packages/decoupler/meta.yaml:12]: Yes, but should the examples that we start with be acronyms?

**Comment by grst** [comment]
Fields to be added: 

 * add version of the package at the time it was added
 * author (list of github handles) 
 * test_command (optional, bash command to install package / execute tests, assuming that you are in a freshly cloned repo)

**Comment by grst** [comment]
Implemented the changes discussed yesterday. Merging, if there's anything else, let's make a follow-up PR.

**Pull request merged**